### PR TITLE
Transfer clifford and cnot-dihedral synthesis functions from quantum_info to synthesis

### PR DESCRIPTION
**Pull request by ShellyGarion**
### Summary

Transfer Clifford and CNOT-Dihedral synthesis functions from `qiskit/quantum_info/synthesis` to `qiskit/synthesis`

### Details and comments

The functions `decompose_clifford` and `decompose_cnotdihedral` will be deprecated, renamed and moved to `qiskit/quantum_info/synthesis`.

### CODE REVIEW

**Comment by ShellyGarion** [comment]
This PR is almost done (only needs release notes). I would like to approve the API before that.

**Review by alexanderivrii — COMMENTED** [review]
Thanks @ShellyGarion. I agree that moving synthesis algorithms for `Clifford` and `CNOTDihedral` into the `qiskit/synthesis` directory makes a lot of sense and makes the code structure a bit cleaner. As far as I can judge, this PR does not do any algorithmic changes (modulo some minor renaming), so overall this looks good to me.

One minor question that I had is whether we still need `clifford_decompose.py` and `cnotdihedral_decompose.py` files in the `quantum_info/synthesis` directory. This is not by any means a suggestion to remove these files, it's only a question if you think we need them in the long run? 

The renamed function names (such as `synch_clifford_ag` instead of `decompose_clifford_ag`) are completely fine, I don't think we really needed to rename these, but I do like the new version slightly better, and maybe we are implicitly converging to a convention that synthesis functions in the `qiskit/synthesis` directory should start with `synth`.

Since now `Clifford` and `CNOTDihedral` synthesis methods (such as `synth_clifford_ag`, `synth_cnotdihedral_full`, etc.) are part of the official documentation, it makes sense to add some release notes.

File: qiskit/quantum_info/synthesis/clifford_decompose.py [hunk 1]
```diff
@@ -12,25 +12,12 @@
...
-    _append_z,
+from qiskit.synthesis.clifford import (
+    synth_clifford_ag,
+    synth_clifford_bm,
+    synth_clifford_greedy,
 )
-from qiskit.quantum_info.operators.symplectic.pauli import Pauli
 
 
 def decompose_clifford(clifford, method=None):
```
Comment (ikkoham) [inline qiskit/quantum_info/synthesis/clifford_decompose.py:23]: Why not transfer this function to `qiskit.synthesis`?

File: qiskit/quantum_info/synthesis/clifford_decompose.py [hunk 2]
```diff
@@ -65,699 +52,12 @@ def decompose_clifford(clifford, method=None):
...
-
-    return decompose_clifford_greedy(clifford)
-
-
-# ---------------------------------------------------------------------
-# Synthesis based on Bravyi & Maslov decomposition
-# ---------------------------------------------------------------------
-
-
-def decompose_clifford_bm(clifford):
```
Comment (ikkoham) [inline qiskit/quantum_info/synthesis/clifford_decompose.py:84]: Removing these public functions are API breaking, so could you add deprecation warning?

**Review by ikkoham — CHANGES_REQUESTED** [review]
Thank you. The direction is nice.
I'm concerned about circular dependencies between `quantum_info` and `synthesis`. I think it would be nice that `synthesis` depends on `quantum_info` and `synthesis` should not depend on `quantum_info`. It is good to be dependent only during the deprecation period. It is transient.
I only wrote comments in one place because I didn't want to write the same again, but there are many similar comments.

And also could you add tests for synthesis and release note?

File: qiskit/quantum_info/synthesis/clifford_decompose.py (hunk 2 above)
Comment (ShellyGarion) [reply qiskit/quantum_info/synthesis/clifford_decompose.py:84]: These functions are internal and not part of the API. The only function appearing in the docs is `decompose_clifford`:
https://qiskit.org/documentation/apidoc/quantum_info.html#synthesis

File: qiskit/quantum_info/synthesis/clifford_decompose.py (hunk 1 above)
Comment (ShellyGarion) [reply qiskit/quantum_info/synthesis/clifford_decompose.py:23]: As you suggested, I will transfer `decompose_clifford` into `qiskit.synthesis` and add a deprecation warning.

**Comment by ShellyGarion** [comment]
Thank you @alexanderivrii and @ikkoham for your reviews!
As suggested, I deprecated the functions `decompose_clifford` and `decompose_cnotdihedral`, and transferred their code to `qiskit.synthesis`. There is no need to deprecate their sub-functions, since they have never appeared in Qiskit API docs.
I also added release notes and some tests.

**Review by ikkoham — COMMENTED** [review]
Thank you! Only one more issue remains.

**Comment by alexanderivrii** [comment]
> Thank you! Only one more issue remains.

@ikkoham, thanks for spotting this. This is a really new feature, and I don't think anyone is really using it yet, so it should probably be ok to just change the documentation without issuing a deprecation warning.

File: qiskit/quantum_info/synthesis/clifford_decompose.py (hunk 2 above)
Comment (ShellyGarion) [reply qiskit/quantum_info/synthesis/clifford_decompose.py:84]: Thanks for catching this! I fixed the documentation accordingly.

**Comment by ikkoham** [comment]
@alexanderivrii Thanks. I understand the situation. I think it's good.

**Review by ikkoham — APPROVED** [review]
Thank you. LGTM!

**Pull request merged**